lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -1279,6 +1282,248 @@ SmallVector<DictionaryAttr> InstanceOp::portAttributes() {
+           << "result type for " << primitivePorts[i].name << " must be "
+           << expectedType << ", but got " << resultType;
+  }
+  return success();
+}
+
+LogicalResult PrimitiveOp::verifySymbolUses(SymbolTableCollection &symbolTable) {
+  Operation *op = *this;
+  auto program = op->getParentOfType<ProgramOp>();
+  Operation *referencedPrimitive =
+      symbolTable.lookupNearestSymbolFrom(program, primitiveNameAttr());
+  if (referencedPrimitive == nullptr)
+    return emitError() << "referencing component: '" << primitiveName()
+                       << "', which does not exist.";
+
+  Operation *shadowedPrimitiveName =
+      symbolTable.lookupNearestSymbolFrom(program, sym_nameAttr());
+  if (shadowedPrimitiveName != nullptr)
+    return emitError() << "instance symbol: '" << instanceName()
+                       << "' is already a symbol for another component.";
+
+  // Verify the referenced component is not instantiating itself.
+  auto parentPrimitive = op->getParentOfType<hw::HWModuleExternOp>();
+  if (parentPrimitive == referencedPrimitive)
+    return emitError() << "recursive instantiation of its parent component: '"
+                       << primitiveName() << "'";
+
+  assert(isa<hw::HWModuleExternOp>(referencedPrimitive) && "Should be a PrimitiveOp.");
+  return verifyPrimitiveOpType(*this, cast<hw::HWModuleExternOp>(referencedPrimitive));
+}
+
+/// Provide meaningful names to the result values of an PrimitiveOp.
+void PrimitiveOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
+  getCellAsmResultNames(setNameFn, *this, this->portNames());
+}
+
+SmallVector<StringRef> PrimitiveOp::portNames() {
+  SmallVector<StringRef> portNames;
+  for (hw::PortInfo port : getReferencedPrimitive().getAllPorts())
+    portNames.push_back(port.name.getValue());
+
+  return portNames;
+}
+
+Direction convertHWDirectionToCalyx(hw::PortDirection direction) {
+  switch (direction) {
+    case hw::PortDirection::INPUT:
+      return Direction::Input;
+    case hw::PortDirection::OUTPUT:
+      return Direction::Output;
+    default:
+      return Direction::Input;
+  }
+}
+
+SmallVector<Direction> PrimitiveOp::portDirections() {
+  SmallVector<Direction> portDirections;
+  for (hw::PortInfo port : getReferencedPrimitive().getAllPorts())
+    portDirections.push_back(convertHWDirectionToCalyx(port.direction));
+  return portDirections;
+}
+
+SmallVector<DictionaryAttr> PrimitiveOp::portAttributes() {
+  SmallVector<DictionaryAttr> portAttributes;
+  for (size_t i = 0; i < getReferencedPrimitive().getAllPorts().size(); i++)
+    portAttributes.push_back(DictionaryAttr());
+  return portAttributes;
+}
+
+/// Parse an parameter list if present. Same format as HW dialect.
+/// module-parameter-list ::= `<` parameter-decl (`,` parameter-decl)* `>`
+/// parameter-decl ::= identifier `:` type
+/// parameter-decl ::= identifier `:` type `=` attribute
+///
+static ParseResult parseParameterList(OpAsmParser &parser,
```

Comment:
I don't want to block progress too much here, but as above, it seems like `parseParameterList`, `checkParameterInContext`, ... are similar (identical?) to what's in the HW dialect. Shared logic could be moved to something like `Support/printerparsers.h` to share parameter-printer/parsing logic - @mikeurbach thoughts?

Reply:
I'm not sure of the best way to do this. I would be totally okay with moving this to some shared location, I was just copying how the MSFT dialect handles ParamDeclAttr parsing. It would make a lot of sense to have both MSFT and Calyx reuse code from the HW dialect though. If I were to put this code in op interfaces instead, what would that look like? Are there any good examples of doing something similar in other dialects?

Reply:
I would also prefer being able to get this work merged and then fixing the duplicate code in a follow up PR, as that change will also be making changes to the MSFT and HW dialects.

Reply:
I think you'd start by adding the shared methods to HWInstanceLike: https://github.com/llvm/circt/blob/d8a8bf1296dbc5e740ba2a5e8e265038c49b195d/include/circt/Dialect/HW/HWOpInterfaces.td#L122

Then, you'd make sure the new PrimitiveOp uses the trait `DeclareOpInterfaceMethods<HWInstanceLike>` in the ODS file.

You'd of course need to implement them somewhere. The best bet is probably to provide a default implementation in the HWInstanceLike interface. That way they could be shared between Calyx Primitive and HW Instance (and probably MSFT Instance).

I don't really have a preference on whether this gets refactored before or after the current PR, but that change is independent. Usually (in the LLVM community), when there's a new feature and a refactor is coming to enable that, we send the refactor first, so the new feature gets to use the newly refactored thing, and both PRs are scoped to just what they are changing.

Reply:
I'm not sure these parser and printers make sense in the HWInstanceLike interface itself, as they aren't tied to a specific HW instance. Looking at the [ViewLikeInterface](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Interfaces/ViewLikeInterface.h#L89) these custom parsers and printers for attributes are included in the .h file, but not declared in the .td.

Reply:
I'm pretty sure this code is duplicated for the use case of parsing/printing HWInstanceLike things right (msft.instance and now calyx.primitive)? That's why I suggested putting it in an OpInterface--the two users should already be depending on a shared interface. Morten's suggestion to put them into their own header/library also makes sense.

But, I think we've actually wanted to move the parsing/printing logic to the HWModuleLike and HWInstanceLike interfaces. Not just parameters, but the whole module/instance parsing logic. That allows us to have the same textual assembly format for the ops implementing the interface, regardless of the dialect.

This isn't a major thing, and it's a bigger ask than what you're doing here, so I agree with your original point--let's leave that for a follow up, and accept the duplication for now.

Reply:
Agreed! Seems like we can open an issue to track this duplication after this is merged and clean things up afterwards